hetio/neo4j.py

```diff
@@ -281,6 +281,171 @@ def construct_dwpc_query(metarels, property='name', join_hint='midpoint', index_
+    degree_query = ',\n'.join(degree_strs)
+
+    using_query = ''
+    # Specify index hint for node lookup
+    if index_hint:
+        using_query = '\n' + textwrap.dedent('''\
+        USING INDEX n0:{source_label}({property})
+        USING INDEX n{length}:{target_label}({property})
+        ''').rstrip().format(
+            property = property,
+            source_label = metarels[0][0],
+            target_label = metarels[-1][1],
+            length = len(metarels)
+        )
+
+    # Specify join hint with node to join on
+    if join_hint is not False:
+        if join_hint is True or join_hint == 'midpoint':
+            join_hint = len(metarels) // 2
+        join_hint = int(join_hint)
+        assert join_hint >= 0
+        assert join_hint <= len(metarels)
+        using_query += "\nUSING JOIN ON n{}".format(join_hint)
+
+    # Unique node constraint (pevent paths with duplicate nodes)
+    if unique_nodes == 'nested':
+        unique_nodes_query = '\nAND ALL (x IN nodes(path) WHERE size(filter(z IN nodes(path) WHERE z = x)) = 1)'
+    elif unique_nodes == 'expanded':
+        pairs = itertools.combinations(range(len(metarels) + 1), 2)
+        unique_nodes_query = format_expanded_clause(pairs)
+    elif unique_nodes == 'labeled' or unique_nodes is True:
+        labels = [metarel[0] for metarel in metarels]
+        labels.append(metarels[-1][1])
+        label_to_nodes = dict()
+        for i, label in enumerate(labels):
+            label_to_nodes.setdefault(label, list()).append(i)
+        pairs = list()
+        for nodes in label_to_nodes.values():
+            pairs.extend(itertools.combinations(nodes, 2))
+        unique_nodes_query = format_expanded_clause(pairs)
+    else:
+        assert unique_nodes is False
+        unique_nodes_query = ''
+
+    # combine cypher fragments into a single query and add PDP logic
+    query = ''
+    if dwpc is not None:
+        query = textwrap.dedent('''\
+            MATCH path = {metapath_query}{using_query}
+            WHERE n0.{property} = {{ source }}
+            AND n{length}.{property} = {{ target }}{unique_nodes_query}
+            WITH
+            [
+            {degree_query}
+            ] AS degrees, path
+            WITH path, reduce(pdp = 1.0, d in degrees| pdp * d ^ -{{ w }}) as PDP
+            RETURN
+            path,
+            PDP,
+            100 * (PDP / {dwpc}) AS PERCENT_OF_DWPC
+            ORDER BY PERCENT_OF_DWPC DESC
+            ''').rstrip().format(
+            metapath_query = metapath_query,
+            using_query = using_query,
+            unique_nodes_query = unique_nodes_query,
+            degree_query = degree_query,
+            length=len(metarels),
+            property=property,
+            dwpc = dwpc)
+    # If the dwpc isn't provided, we'll have to calculate it before the PDP.
+    # Doing so roughly doubles the query execution time, as it effectively
+    # runs the query twice returning different degrees of aggregation.
```

Comment:
Let's open a stack overflow issue with the `neo4j` tag on whether we can do a sum aggregation and then combine that with the pre-aggregated result table all in Cypher.